fractions/keycloak/src/main/java/org/wildfly/swarm/keycloak/runtime/SecuredArchivePreparer.java

```diff
@@ -143,10 +165,61 @@ private Asset createAsset(InputStream in) throws IOException {
+        if (keycloakMultitenancyPaths != null && !keycloakMultitenancyPaths.isEmpty()) {
+            // Prepare a relative paths to KC path deployments map.
+            Map<String, KeycloakDeployment> pathDeployments = new HashMap<>();
+            for (Map.Entry<String, String> entry : keycloakMultitenancyPaths.entrySet()) {
+                InputStream is = getKeycloakJsonStream(entry.getValue());
+                if (is == null) {
+                    LOG.warn(String.format(
+                        "Unable to support the multitenancy due to the '%s' being not available", entry.getValue()
+                    ));
+                } else {
+                    pathDeployments.put(entry.getKey(), KeycloakDeploymentBuilder.build(is));
+                }
+            }
+            try {
+                // Statically initialize KeycloakAdapterConfigResolver given that
+                // KeyCloak can only load it as opposed to accepting the prepared instance.
+                Module module = Module.getBootModuleLoader().loadModule("org.wildfly.swarm.keycloak:deployment");
+                final String resolverClassName = "org.wildfly.swarm.keycloak.deployment.KeycloakAdapterConfigResolver";
+                Class<?> resolverClass = module.getClassLoader().loadClass(resolverClassName);
+                Method setPathDeployments = resolverClass.getDeclaredMethod("setPathDeployments", Map.class);
+                setPathDeployments.invoke(null, pathDeployments);
+
+                // Set "keycloak.config.resolver" context parameter for KC be able to load it
+                WebXmlAsset webXmlAsset = null;
+                Node node = archive.as(JARArchive.class).get("WEB-INF/web.xml");
+                if (node == null) {
+                    webXmlAsset = new WebXmlAsset();
+                    archive.as(JARArchive.class).add(webXmlAsset);
+                } else {
+                    Asset asset = node.getAsset();
+                    if (!(asset instanceof WebXmlAsset)) {
+                        webXmlAsset = new WebXmlAsset(asset.openStream());
+                        archive.as(JARArchive.class).add(webXmlAsset);
+                    } else {
+                        webXmlAsset = (WebXmlAsset) asset;
+                    }
+                }
+                webXmlAsset.setContextParam("keycloak.config.resolver", resolverClassName);
+            } catch (Throwable e) {
+                LOG.warn(String.format(
+                        "KeycloakAdapterConfigResolver can not be set up, multitenancy will not be supported", e
+                    ));
+            }
+        }
+    }
+
     private final Archive<?> archive;
 
-    @AttributeDocumentation("Path to keycloak.json configuration")
+    @AttributeDocumentation("Path to Keycloak adapter configuration")
     @Configurable("swarm.keycloak.json.path")
     String keycloakJsonPath;
 
+    @AttributeDocumentation("Keycloak multitenancy paths configuration")
```

Comment:
Not very descriptive, also not sure how extensible it is, but whatevz :-)

Reply:
Now it should be more extensible :-). Originally I had it under 'swarm.keycloak.multitenancy' - but then thought that if if were to support the query or header or claim based dynamic config resolution then it would not fir all under 'swarm.keycloak.multitenancy'. Hence it is now ''swarm.keycloak.multitenancy.paths' for the path-based resolution, later if needed we can have "swarm.keycloak.multitenancy.queries", ".headers", etc, example: "headers: headerA" where the value of the headerA will be used to select a deployment from the 'headersDeployments' etc. Similarly for the claim-based resolution...

Let me try to update the doc a bit

Reply:
Yea that's what I imagine, but no idea how these multiple config items will compose :-) But that's a bit premature I guess.

Reply:
They can probably be composed like this: check the path based deployments, next - queries, next headers. It is quite likely that for the cases where the dynamic resolution will require some more involved checks, say select a given deployment only if a given path and a given header and may be two JWT claims have some values, all at the same time, the users will just do it themselves.
The goal for this minor enhancement is to support few common mainstream cases and let users do it with just some basic config...
It would be good if we start getting some enhancement requests around it later on :-)

Reply:
Ah that makes sense. Thanks!